shell/platform/embedder/embedder_semantics_update.cc

```diff
@@ -175,4 +193,61 @@ void EmbedderSemanticsUpdate2::AddAction(
   });
 }
 
+std::pair<size_t, const FlutterStringAttribute**>
+EmbedderSemanticsUpdate2::CreateStringAttributes(
```

Comment:
Still pondering this method. Maybe it's worth doing a quick brain dump of my disorganised thoughts to see if something coherent comes out (probably not):
* This method is both mutating state but returning (part of) the same state (`node_string_attributes_`).
* `std::pair` usually just means there are two return values, but you already have both those values on the same struct (`node_string_attributes_`) so why not just return a const ref to that?
* Alternatively, could name it `Update[blah]` and just use the underlying vectors in the method above?

My list of stuff remaining to understand while I review this:
* The same method is being called for 5 different sets of attributes, but also mutating the same underlying state. Why is that state at the EmbedderSemanticsUpdate2 scope? It seems like it's being overwritten after each call, in which case it's temporary, so why not make this local within this method, in which case you'd return the vector rather than a const ref to it?
* Why the shared_ptr wrangling?
* Think about object lifetimes, in particular of the string attributes, since it doesn't look like we're making copies of the contents.

Reply:
> * `std::pair` usually just means there are two return values, but you already have both those values on the same struct (`node_string_attributes_`) so why not just return a const ref to that?

That works too, in fact my original implementation did that. I switched to the current implementation so that all the gnarly memory management is encapsulated by `CreateStringAttributes`. This also makes the "don't allocate if there's no string attributes" case a bit cleaner.

> * The same method is being called for 5 different sets of attributes, but also mutating the same underlying state. Why is that state at the EmbedderSemanticsUpdate2 scope? It seems like it's being overwritten after each call, in which case it's temporary, so why not make this local within this method, in which case you'd return the vector rather than a const ref to it?

We're _appending_ memory allocations for bookkeeping purposes so that they are deallocated when `EmbedderSemanticsUpdate2` is destroyed. Previously created embedder objects are not mutated.

The `EmbedderSemanticsUpdate2` exists for the duration of the `update_semantics_callback2` callback:

https://github.com/flutter/engine/blob/9fae689d10a9fc7615b9b185172b1bed5c8d40a1/shell/platform/embedder/embedder.cc#L1565-L1576

> * Think about object lifetimes, in particular of the string attributes, since it doesn't look like we're making copies of the contents.

All of the arrays/structs/strings become invalid once the update semantics callback ends. It is the responsibility of the embedder to copy any data it needs to survive past the update semantics callback. This mirrors the existing update semantics behavior.

Here's where the embedders copy values from the update semantics callback to persist them after the callback ends:

https://github.com/flutter/engine/blob/9fae689d10a9fc7615b9b185172b1bed5c8d40a1/shell/platform/common/accessibility_bridge.cc#L574-L623

> * Why the shared_ptr wrangling?

The engine uses an inheritance hierarchy for string attributes. There's a `StringAttribute` base class that's subclassed by `SpellOutStringAttribute` and `LocaleStringAttribute`. The shared_ptr wrangling is used to access the subclass's members. Let me know if there's a better way to do this.

Reply:
Let me know if you'd like to chat about this on a call. This stuff is dense, I'd be happy to talk through it :)

Reply:
Thanks -- sorry to be clear, those were notes of what I still needed to review as I hadn't done anything but done a superficial glance at the method and noticed a few things that are normally review "dig deeper as to why" code smells. (pair, shared_ptr, being both a mutator and a getter at the same time)

After reading through, I think it's pretty clear to the reader that that the three vectors are there purely for lifetime management of the temporary objects being returned to the embedder API so I don't think we need any more detailed comments here, but I do think it's worth adding doc comments to the three vectors for future readers. (just one comment above the three of them maybe).

I didn't look into why our internal APIs in string_attribute.h rely on shared_ptr, but for the purposes of this patch that's irrelevant and we're stuck with em either way. I'll take it on faith that they're necessary in there.

Thanks for writing up the notes here!